tools/benchcomp/benchcomp/entry/run.py

```diff
@@ -0,0 +1,133 @@
+# `suite.yaml`.Parsers are python submodules of benchcomp.parsers; the
+# configuration file describes which parser to use for each benchmark suite.
+
+
+import dataclasses
+import importlib
+import logging
+import os
+import pathlib
+import shutil
+import subprocess
+import uuid
+
+import yaml
+
+import benchcomp
+
+
+@dataclasses.dataclass
+class _SingleInvocation:
+    """Run and parse the result of a single suite x variant"""
+
+    suite_id: str
+    variant_id: str
+
+    parser: str
+
+    suite_yaml_out_dir: pathlib.Path
+
+    command_line: str
+    directory: pathlib.Path
+
+    env: dict = dataclasses.field(default_factory=dict)
+    timeout: int = None
+    memount: int = None
+    patches: list = dataclasses.field(default_factory=list)
+
+    def __post_init__(self):
+        self.working_copy: pathlib.Path = pathlib.Path(
+            f"/tmp/benchcomp/suites/{uuid.uuid4()}")
+
+    def __call__(self):
+        env = dict(os.environ)
+        env.update(self.env)
+
+        shutil.copytree(self.directory, self.working_copy)
+
+        try:
+            subprocess.run(
+                self.command_line, shell=True, env=env, cwd=self.working_copy,
+                check=True)
+        except subprocess.CalledProcessError as exc:
+            logging.warning(
+                "Invocation of suite %s with variant %s exited with code %d",
+                self.suite_id, self.variant_id, exc.returncode)
+            return
+        except (OSError, subprocess.SubprocessError):
+            logging.error(
+                "Invocation of suite %s with variant %s failed", self.suite_id,
+                self.variant_id)
+            return
+
+        parser_mod_name = f"benchcomp.parsers.{self.parser}"
+        parser = importlib.import_module(parser_mod_name)
+        suite = parser.main(self.working_copy)
+
+        suite["suite_id"] = self.suite_id
+        suite["variant_id"] = self.variant_id
+
+        out_file = f"{self.suite_id}@{self.variant_id}_suite.yaml"
+        with open(
+                self.suite_yaml_out_dir / out_file, "w",
+                encoding="utf-8") as handle:
+            yaml.dump(suite, handle, default_flow_style=False)
+
+
+@dataclasses.dataclass
+class _Run:
+    """Run all suite x variant combinations, write results to a directory"""
+
+    config: benchcomp.ConfigFile
+    out_prefix: pathlib.Path
+    out_dir: str
+    out_symlink: str
+    result: dict = None
+
+    def __call__(self):
+        out_path = (self.out_prefix / self.out_dir)
+        out_path.mkdir(parents=True)
+
+        for suite_id, suite in self.config["run"]["suites"].items():
+            for variant_id in suite["variants"]:
+                variant = self.config["variants"][variant_id]
+                config = dict(variant).pop("config")
+                invoke = _SingleInvocation(
+                    suite_id, variant_id,
+                    suite["parser"]["module"],
```

Comment:
What does the "module" key represent?

Reply:
This should be the name of a python module under `benchcomp.parsers`. The module must have a main method, etc. which is currently documented in `benchcomp/parsers/README.md`.  Currently the only such module is `benchcomp.parsers.test` which is used by the test suite, but I'll be implementing a real parser shortly. Each kind of suite will have its own parser in that directory.

As future work, a config file might also specify `"executable"` instead of `"module"`. This would be treated as a script that will be run in the suite directory, and is expected to print the (JSON or YAML) result to stdout. The python module doesn't print anything, it returns a dict instead.

I realise this needs to be documented, but if you don't mind I'd prefer to leave it to a future PR. I have some code that interleaves the documentation with the config file schema itself (and similarly for the schemas of suite.yaml and result.yaml), which will allow benchcomp to automatically render nicely-formatted documentation that users can read without reading the code.